text/0000-symbol-name-mangling-v2.md

```diff
@@ -0,0 +1,799 @@
+
+- Using UTF-8 instead of Punycode would make mangled strings containing non-ASCII identifiers a bit more human-readable. For demangled strings, there would be no difference.
+
+- Punycode support is non-optional since some platforms only allow a very limited character set for symbol names. Thus, we would be using UTF-8 on some platforms and Punycode on others, making it harder to predict what a symbol name for a given item looks like.
+
+- Punycode encoding and decoding is more runtime effort for the mangler and demangler.
+
+- Once a demangler supports Punycode, it is not much effort to support both encodings. The `u` identifier suffix tells the demangler whether it's Punycode. Otherwise it can just assume UTF-8 which already subsumes ASCII.
+
+### Re-use <disambiguator> for crate disambiguator
+
+The RFC currently proposes to represent crate-ids as an `<identifier>` of the form `<name>_<disambiguator>`. However, the `<identifier>` production already supports disambiguation via its `<disambiguator>` component. The crate disambiguator could be encoded into an disambiguation index.
+
+# Appendix A - Suggested Demangling
+
+This RFC suggests that names are demangled to a form that matches Rust syntax as it is used in source code, compiler error messages and `rustdoc`:
+
+ - Path components should be separated by `::`.
+
+ - If the path root is a `<crate-id>` it should be printed as the crate name. If the context requires it for correctness, the crate disambiguator should be printed too, as in, for example, `std[a0b1c2d3]::collections::HashMap`. In this case `a0b1c2d3` would be the disambiguator. Usually, the disambiguator can be omitted for better readability.
+
+ - If the path root is a trait impl, it should be printed as `<SelfType as Trait>`, like the compiler does in error messages.
+
+ - The list of generic arguments should be demangled as `<T1, T2, T3>`.
+
+ - Identifiers and trait impl path roots can have a numeric disambiguator (the `<disambiguator>` production). The syntactic version of the numeric disambiguator maps to a numeric index. If the disambiguator is not present, this index is 0. If it is of the form `s_` then the index is 1. If it is of the form `s<base-62-digit>_` then the index is `<base-62-digit> + 2`. The suggested demangling of a disambiguator is `[<index>]`. However, for better readability, these disambiguators should usually be omitted in the demangling altogether. Disambiguators with index zero can always be omitted.
+
+ The exception here are closures. Since these do not have a name, the disambiguator is the only thing identifying them. The suggested demangling for closures is thus `{closure}[<index>]`.
+
+ - In a lossless demangling, identifiers from the value namespace should be marked with a `'` suffix in order to avoid conflicts with identifiers from the type namespace. In a user-facing demangling, where such conflicts are acceptable, the suffix can be omitted.
```

Comment:
Wouldn't that include all the statics and functions? Seems a bit excessive.

Reply:
It does, but I don't think there's a way around it. Otherwise you get conflicts for examples like:
```rust
fn foo() {
    fn bar() {}
}

mod foo {
    fn bar() {}
}
```
Note though that this is only for "lossless" demanglings. For most user-facing demanglings, like in debuggers or backtraces, the suffix can just be omitted. I suggest that demanglers support `lossless` or `verbose` option that is usually set to `false`.